### PR TITLE
Master

### PR DESCRIPTION
**Pull request by coffeehc**
1.add response stream out
2.add render method GetFinalOptions()
3.add render method GetCompiledCharset()

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Compiled Object files, Static and Dynamic libs (Shared Objects)
 *.o
 *.a
+/.idea
```
Comment (unrolled) [inline .gitignore]: Remove from the project gitignore. Things like this are usually better suited in your global ignore file.

File: engine.go [hunk 2]
```diff
@@ -59,12 +60,42 @@ type XML struct {
...
+	Head
+}
+
 // Write outputs the header content.
 func (h Head) Write(w http.ResponseWriter) {
 	w.Header().Set(ContentType, h.ContentType)
 	w.WriteHeader(h.Status)
 }
 
+//Remder a stream response.
```
Comment (unrolled) [inline engine.go]: s/Remder/Render

Also add a space between the slashes and your comment.

File: engine.go [hunk 3]
```diff
@@ -59,12 +60,42 @@ type XML struct {
...
+func (s Stream)Render(w http.ResponseWriter, v interface{}) error {
```
Comment (unrolled) [inline engine.go]: Instead of passing in an interface, could you pass in an `io.Reader`object?

File: render.go [hunk 4]
```diff
@@ -125,6 +126,14 @@ func New(options ...Options) *Render {
 
 	return &r
 }
+//get finally options
+func (r *Render)GetFinallyOptions() Options{
```
Comment (unrolled) [inline render.go]: This isn't used in your code... and I've intentionally kept the options internal. Please remove.

File: render.go [hunk 5]
```diff
@@ -125,6 +126,14 @@ func New(options ...Options) *Render {
...
+	return r.opt
+}
+//get compiledCharset for extend response
+func (r *Render)GetCompiledCharset() string{
```
Comment (unrolled) [inline render.go]: Same goes for this function.

**Comment by unrolled** [comment]
Thanks for the PR! 

Can you give me a bit of information around what you are trying to add here? Are you streaming specialized content? As an FYI there is a streaming option for JSON already. Also, could you add a few tests for you new function? 

Thanks!

**Comment by coffeehc** [comment]
Sometimes I need to look through custom output encoded binary content ( such as content pb ) , it is necessary to increase the stream method.

In addition, I own the code which encapsulates again,see[transport.go](https://github.com/coffeehc/web/blob/master/transport.go)

**Pull request closed without merging**